### ISSUE TITLE
Flash messages not preserved over redirect

### ISSUE THREAD

**Issue by ChristianVermeulen** [symfony/symfony#50952]
### Symfony version(s) affected

6.3.1

### Description

When adding flashmessages they are not preserved over a redirect. If the response is returned directly they show, but when returning a RedirectResponse the flash is part of the session in the 302 response, but no longer in the final destination.

### How to reproduce

```php

public function __construct(
    private readonly RequestStack $requestStack
) {}

public function someAction(Request $request): Response
{
    # method 1
    $request->getSession()->getFlashBag()->add('success', 'some message');

    # method 2
    $this->requestStack->getSession()->getFlashBag()->add('success', 'some message');

    return new RedirectResponse('/some-page');
}
```

Using either method, `/some-page` no langer has any flash messages in the session. When returning a normal `Response($twig->render('view.html.twig'))` the flash messages are still there. When inspecting redirect response in the profiler, it does hold the flash.

![Screenshot 2023-07-12 at 12 42 24@2x](https://github.com/symfony/symfony/assets/1062751/7423ed06-c6f9-4062-8339-a81f47ddc779)

**Comment by MatTheCat** [comment]
I cannot reproduce on a personal project: flash messages still are in the session after a redirect. Could you provide a repository of the simplest application where the issue is present?

**Comment by 94noni** [comment]
if you use the this->addFlash from the abstract controller does it work?
Cant reproduce neither on latest version but i use the shortcut

**Comment by ChristianVermeulen** [comment]
> I cannot reproduce on a personal project: flash messages still are in the session after a redirect. Could you provide a repository of the simplest application where the issue is present?

I would have to build one from scratch, we have a rather large application 😅


> if you use the this->addFlash from the abstract controller does it work? Cant reproduce neither on latest version but i use the shortcut

`$this->addFlash()` is just a shortcut to `$this->requestStack->getSession()->getFlashBag()->add('success', 'some message');`.

We do not extend the abstract controllers because we mostly only use the router and translator so we create our own services to save on resources.

**Comment by xabbuh** [comment]
If you want to debug yourself what is going wrong in your application, I would start with checking that the session the flash message is written to is the same (i.e. the ids are the same) as the one flash messages are read from. Maybe sessions are not written properly in your application.

**Comment by xabbuh** [comment]
@ChristianVermeulen Do you have some more information for us?

**Comment by jmperruchini** [comment]
hi, i have the same problem, I just updated my 5.4 application, PHP 7.4 to symfony 6.3 and PHP 8.2, everything works fine except for one thing that I can't explain to myself! flash messages not working after a redirect! if I do an addFlash on an edit function before the submit it works, but after the if($form->isSubmitted() && $form->isValid()) and just before the redirectToroute, nothing appears!!! has this happened to anyone?
I changed the session configuration in framework.yaml to see what is happening and I entrusted the management to symfony instead of php,

    session:
#        handler_id: null
        handler_id: 'session.handler.native_file'
        save_path: '%kernel.project_dir%/var/sessions/%kernel.environment%'
        cookie_secure: auto
        cookie_samesite: lax
        storage_factory_id: session.storage.factory.native


 I have 2 sessions created, one of which is empty, is this normal?

**Comment by jmperruchini** [comment]
I found the solution, my firewall was set to lazy=true, it works if I set it to false, I have services that handle sessions, this was probably interfering with the management of sessions in lazy mode

**Comment by xabbuh** [comment]
I am going to close here for now due to the lack of feedback. Please let us know when you have more information and we can consider to reopen.

**Comment by ChristianVermeulen** [comment]
Appologies for the late reply, I was on annual leave for the past 3 weeks.

> If you want to debug yourself what is going wrong in your application, I would start with checking that the session the flash message is written to is the same (i.e. the ids are the same) as the one flash messages are read from. Maybe sessions are not written properly in your application.

@xabbuh I have checked this and the session id remains the same. I have tested it with getting the session from the `Request` as well as the `RequestStack` at writing the flash message and reading the flashes in the template. Always the same session id.



> I found the solution, my firewall was set to lazy=true, it works if I set it to false, I have services that handle sessions, this was probably interfering with the management of sessions in lazy mode

@jmperruchini I have tried this as well, but unfortunately it did not do anything, i'm still not getting them after a redirect :-(.

**Comment by ChristianVermeulen** [comment]
Ok, so after trying to set up a small symfony app to test and share, I noticed it was working in a brand new app. So I want back into our code and dumped the bag and id's in several different places.

After a lot debugging I did a search on `SessionBagInterface->clear()` and I'll be damned! Someone had deemed it necessary to clear the entire bag in a request listener 🙈😅. That was a b*** to find, but we've got it!

Issue can be set to resolved now 😄
